git-worktree/src/lib.rs

```diff
@@ -1 +1,126 @@
 #![forbid(unsafe_code, rust_2018_idioms)]
+//! Git Worktree
+
+use git_hash::oid;
+use git_object::bstr::ByteSlice;
+use quick_error::quick_error;
+use std::convert::TryFrom;
+use std::fs;
+use std::fs::create_dir_all;
+use std::path::Path;
+use std::time::Duration;
+
+#[cfg(unix)]
+use std::os::unix::fs::PermissionsExt;
+
+quick_error! {
+    #[derive(Debug)]
+    pub enum Error {
+        Utf8Error(err: git_object::bstr::Utf8Error) {
+            from()
+            display("Could not convert path to UTF8: {}", err)
+        }
+        TimeError(err: std::time::SystemTimeError) {
+            from()
+            display("Could not read file time in proper format: {}", err)
+        }
+        ToU32Error(err: std::num::TryFromIntError) {
+            from()
+            display("Could not convert seconds to u32: {}", err)
+        }
+        IoError(err: std::io::Error) {
+            from()
+            display("IO error while writing blob or reading file metadata or changing filetype: {}", err)
+        }
+        FindOidError(oid: git_hash::ObjectId, path: std::path::PathBuf) {
+            display("unable to read sha1 file of {} ({})", path.display(), oid.to_hex())
+        }
+    }
+}
+
+/// Copy index to `path`
+pub fn copy_index<P, Find>(state: &mut git_index::State, path: P, mut find: Find, opts: Options) -> Result<(), Error>
+where
+    P: AsRef<Path>,
+    Find: for<'a> FnMut(&oid, &'a mut Vec<u8>) -> Option<git_object::BlobRef<'a>>,
+{
+    let path = path.as_ref();
+    let mut buf = Vec::new();
+    let mut entry_time = Vec::new(); // Entries whose timestamps have to be updated
+    for (i, entry) in state.entries().iter().enumerate() {
+        if entry.flags.contains(git_index::entry::Flags::SKIP_WORKTREE) {
+            continue;
+        }
+        let entry_path = entry.path(state).to_path()?;
+        let dest = path.join(entry_path);
+        create_dir_all(dest.parent().expect("entry paths are never empty"))?;
+        match entry.mode {
+            git_index::entry::Mode::FILE | git_index::entry::Mode::FILE_EXECUTABLE => {
+                let obj = find(&entry.id, &mut buf).ok_or_else(|| Error::FindOidError(entry.id, path.to_path_buf()))?;
+                std::fs::write(&dest, obj.data)?;
+                if entry.mode == git_index::entry::Mode::FILE_EXECUTABLE {
+                    #[cfg(unix)]
+                    fs::set_permissions(&dest, fs::Permissions::from_mode(0o777))?;
+                }
+                let met = std::fs::symlink_metadata(&dest)?;
+                let ctime = met
+                    .created()
+                    .map_or(Ok(Duration::from_secs(0)), |x| x.duration_since(std::time::UNIX_EPOCH));
```

Comment:
This should be `let ctime = met.created()?.duration_since(std::time::UNIX_EPOCH)?;
`

Reply:
This will fail on Unix as unix does not store `ctime`. My implementation just sets the `ctime` to `0` if it does not exist as you had previously suggested.

Reply:
Done: https://github.com/Byron/gitoxide/pull/315/commits/a4b880cf17665b61e3f7f193de57704b1db5318f